### PR TITLE
Create unit_tests.yml

### PR DESCRIPTION
**Pull request by bear-in-the-air**

### CODE REVIEW

File: .github/workflows/unit_tests.yml [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    # See: https://docs.github.com/en/free-pro-team@latest/actions/learn-github-actions/managing-complex-workflows#using-a-build-matrix
+    runs-on: self-hosted
+
+    steps:
+    - uses: actions/checkout@v3
+
+    - name: Configure CMake
+      # Configure CMake in a 'build' subdirectory. `CMAKE_BUILD_TYPE` is only required if you are using a single-configuration generator such as make.
+      # See https://cmake.org/cmake/help/latest/variable/CMAKE_BUILD_TYPE.html?highlight=cmake_build_type
+      run: cmake -B ${{github.workspace}}/build -DCMAKE_BUILD_TYPE=${{env.BUILD_TYPE}}
```
Comment (Jambls) [inline .github/workflows/unit_tests.yml:24]: Don't you want to use matthew's script for this?

Comment (bear-in-the-air) [reply .github/workflows/unit_tests.yml:24]: Yes, however I started with that as Matthew hadn't pushed his scripts yet. Really this shouldn't have a PR yet, but I need it to test everything is working correctly.

File: .github/workflows/unit_tests.yml [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+  build:
+    # This script uses linux specific commands, so is unlikely to work on a windows runner.
+    runs-on: self-hosted
+
+    steps:
+    - uses: actions/checkout@v3
+
+    - name: Configure CMake
+      working-directory: ${{github.workspace}}
+      run: setup.sh
```
Comment (bear-in-the-air) [inline .github/workflows/unit_tests.yml:21]: This script contains "echo "To build all and run tests, do the following: $ cd build && make && cd unit_tests && ctest"" Which is a little out of place for the CI. Is it bad practice to have it shouting that in to the void?

**Comment by bear-in-the-air** [comment]
Need to check with @MHCooke that I'm using the static analysis script correctly.

File: .github/workflows/unit_tests.yml (hunk 2 above)
Comment (Jambls) [reply .github/workflows/unit_tests.yml]: is it shouting into the void? aren't the logs stored?

Comment (bear-in-the-air) [reply .github/workflows/unit_tests.yml]: It is storing it as a log file, but you can't exactly run the command from the log file. I'm mostly just wondering if that's bad practice or we'll put up with it.

Comment (MHCooke) [reply .github/workflows/unit_tests.yml]: I don't see a problem with it, imo it's ideal to have the CI using scripts as close as possible to what an end user would be. That means that if our scripts start breaking for some reason we will know asap.

If you are really bothered you could add `> /dev/null` to send stout to null

**Comment by MHCooke** [comment]
image

Could we get it so that each one of these has a useful name, and that there are not duplicates for pull requests and pushes?

Also, you don't need to build the output to run the static analysis or code formatting checks, so you should remove those stages from those flows

**Review by Jambls — APPROVED** [review]

**Comment by MHCooke** [comment]
Done some fiddling with the actions scripts and it looks like you can configure this to have multuple badges with one file. See #48

**Pull request merged**